api_definition_test.go

```diff
@@ -145,6 +181,17 @@ func TestNotVersioned(t *testing.T) {
 	}
 }
 
+func TestDefaultVersion(t *testing.T) {
```

Comment:
This can't be really counted as a real test because it does not show that request was actually sent to the right version. Also, it does not go through HTTP pipeline. Also, I see that you copy pasted API definition, and a lot of the stuff in it does not use at all. There is a better way of writing tests for the gateway, see `buildAndLoadAPI` function usage. But basic idea that your test will look like:

```
func TestDefaultVersion(t *testing.T) {
	config.Global.ListenAddress = "127.0.0.1"

	ln, _ := generateListener(0)
	baseURL := "http://" + ln.Addr().String()
	listen(ln, nil, nil)
	defer func() {
		config.Global.ListenAddress = ""
		ln.Close()
	}()

        buildAndLoadAPI(func(spec *APISpec) {
           v1 := apidef.VirtualMeta{Name: "v1"}
           v1.Paths.WhiteList = []string{"/"}

           v2 := apidef.VirtualMeta{Name: "v2"}
           v2.Paths.WhiteList = []string{"/get"}
           
           spec.VersionData.Location = "url-param"
           spec.VersionData.Key = "v"

           spec.VersionData.Versions["v1"] = v1
           spec.VersionData.Versions["v2"] = v2
           spec.VersionData.DefaultVersion = "v2"
        })
        
        testCases := [...]struct {
		url     string
		code int
	}{
		{"/", 404}, // Not whitelisted for default v2
		{"/get", 200},  // Whitelisted for default v2
		{"/?v=v1", 200}, // Allowed for v1
		{"/?v=v1", 404}, // Not allowed for v1
	}

        for _, tc := range testCases {
          resp, err := http.Get(baseURL + tc.url)
          if err !=nil {
            t.Fatal(err, tc)
          }

           if resp.StatusCode != tc.code {
             t.Error(tc.url, resp.StatusCode, tc.code)
           }
        }
}
```

Reply:
Excellent, I was struggling to see how to do a better test like this. Thanks.

Reply:
Im not quite sure what the VirtualMeta has to do with this as thats related to virtual paths?

Reply:
Worked it out nvm